// @notenoughupdates/discord-akairo/dist/src/struct/listeners/Listener.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListenerOptions = exports.ListenerType = exports.Listener = void 0;
const node_events_1 = __importDefault(require("node:events"));
const zod_1 = require("zod");
const Util_js_1 = require("../../util/Util.js");
const AkairoModule_js_1 = require("../AkairoModule.js");
class Listener extends AkairoModule_js_1.AkairoModule {
    constructor(id, options) {
        exports.ListenerOptions.parse(options);
        const { category, emitter, event, type = "on" } = options;
        super(id, { category });
        this.emitter = emitter;
        this.event = event;
        this.type = type;
    }
}
exports.Listener = Listener;
(0, Util_js_1.patchAbstract)(Listener, "exec");
exports.ListenerType = zod_1.z.enum(["on", "once", "prependListener", "prependOnceListener"]);
exports.ListenerOptions = AkairoModule_js_1.AkairoModuleOptions.extend({
    emitter: zod_1.z.union([zod_1.z.string(), zod_1.z.instanceof(node_events_1.default)]),
    event: zod_1.z.string(),
    type: exports.ListenerType.optional()
}).passthrough();
