// @notenoughupdates/discord-akairo/dist/src/util/AkairoMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AkairoMessage = void 0;
const discord_js_1 = require("discord.js");
class AkairoMessage extends discord_js_1.Base {
    constructor(client, interaction) {
        super(client);
        this.author = interaction.user;
        this.applicationId = interaction.applicationId;
        this.channelId = interaction.channelId;
        this.content = interaction.toString();
        this.createdTimestamp = interaction.createdTimestamp;
        this.guildId = interaction.guildId;
        this.id = interaction.id;
        this.interaction = interaction;
        this.member = interaction.member;
        this.partial = false;
    }
    get channel() {
        return this.interaction.channel;
    }
    get cleanContent() {
        return this.content != null ? (0, discord_js_1.cleanContent)(this.content, this.channel) : null;
    }
    get guild() {
        return this.interaction.guild;
    }
    get createdAt() {
        return this.interaction.createdAt;
    }
    get url() {
        return this.interaction.ephemeral
            ? null
            : `https://discord.com/channels/${this.guild ? this.guild.id : "@me"}/${this.channel?.id}/${this.id}`;
    }
    inGuild() {
        return this.interaction.inCachedGuild();
    }
    delete() {
        return this.interaction.deleteReply();
    }
    reply(options) {
        return this.util.reply(options);
    }
}
exports.AkairoMessage = AkairoMessage;
