// @notenoughupdates/discord-akairo/dist/src/struct/commands/Flag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlagType = exports.Flag = void 0;
class Flag {
    constructor(type, data = {}) {
        this.type = type;
        Object.assign(this, data);
    }
    static cancel() {
        return new Flag(FlagType.Cancel);
    }
    static timeout(time) {
        return new Flag(FlagType.Timeout, { time });
    }
    static retry(message) {
        return new Flag(FlagType.Retry, { message });
    }
    static fail(value) {
        return new Flag(FlagType.Fail, { value });
    }
    static continue(command, ignore = false, rest = null) {
        return new Flag(FlagType.Continue, { command, ignore, rest });
    }
    static is(value, type) {
        return value instanceof Flag && value.type === type;
    }
}
exports.Flag = Flag;
var FlagType;
(function (FlagType) {
    FlagType["Cancel"] = "cancel";
    FlagType["Timeout"] = "timeout";
    FlagType["Retry"] = "retry";
    FlagType["Fail"] = "fail";
    FlagType["Continue"] = "continue";
})(FlagType = exports.FlagType || (exports.FlagType = {}));
